### PR TITLE
Update to Rust 2018, syn/quote/proc-macro2 1.0

### LINKED ISSUE #164: Upgrade syn/quote/proc-macro2 to 1.0
Assuming #154 decides to bump SNAFU's Rust requirement to 1.31, we'll be able to upgrade our syn, quote, and proc-macro2 dependencies to 1.0, which also require Rust 1.31.

https://github.com/dtolnay/syn/releases/tag/1.0.0

Some potential benefits mentioned in #145: "making more use of the powers given by the new versions, maybe mostly in quote. For example all of the repetition things mentioned are issues we encountered."

Potentially related, also from #145: "I've had a mild sense that snafu-derive's lib.rs is simply too big at this point and could benefit from some reorganization. Perhaps reviewing the code during such reorganization would be a good time."

### PR DESCRIPTION
**Pull request by tjkirch**
Decision made in #154.
Fixes #164.

This takes advantage of all of the improvements in syn/quote/proc-macro2 1.0 that I could find in the code, and switches to some Rust 2018 idioms, but more could probably be done with more familiarity or research.

Certainly some of the changes in my commits are debatable, for example removing vs. replacing the `rust_1_30` feature.  We don't have anything conditional on version after this, so I opted for the simpler result of removing it.

Unit and compatibility tests pass.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -50,17 +47,17 @@ rust_1_30 = ["snafu-derive/rust_1_30", "doc-comment"]
...
-  "compatibility-tests/v1_30",
+  "compatibility-tests/v1_31",
   "compatibility-tests/futures-0.1",
   "compatibility-tests/futures",
   "compatibility-tests/without-backtrace",
 ]
 
 [dependencies]
 snafu-derive = { path = "snafu-derive", version = "0.5.0" }
+doc-comment = "0.3.1"
```
Comment (shepmaster) [inline Cargo.toml]: Let's keep `default-features = false`. This helps minimize compile time where possible.

File: compatibility-tests/v1_31/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,10 @@
+[package]
+name = "v1_31"
+version = "0.1.0"
+authors = ["Jake Goulding <jake.goulding@gmail.com>"]
+
+[dependencies]
+snafu = { path = "../..", default-features = false, features = ["backtraces"] }
+
+# These versions are the last versions to support Rust 1.31
+# <none yet>
```
Comment (shepmaster) [inline compatibility-tests/v1_31/Cargo.toml]: Heh, looks like we spoke too soon. Evidently the backtrace crate needs to be pinned to an earlier version here.

**Comment by tjkirch** [comment]
Trying to figure out why the 1.31 test passes locally but fails in CI...

File: compatibility-tests/v1_31/Cargo.toml (hunk 2 above)
Comment (tjkirch) [reply compatibility-tests/v1_31/Cargo.toml]: Yep!  Sorry, forgot to clean my compatibility-test Cargo.lock files.  I wonder if I can automate that somehow...

I found it requires 0.3.35 at most.  backtrace-sys doesn't have to be pinned.  I just pushed that fix.

File: Cargo.toml (hunk 1 above)
Comment (tjkirch) [reply Cargo.toml]: Sorry about that; just pushed the fix in the 1.30->1.31 commit.

**Pull request merged**